drivers/spi/spi_rpi_pico_pio.c

```diff
@@ -258,28 +543,228 @@ static void spi_pico_pio_txrx(const struct device *dev)
 			/* Send 0 in the case of read only operation */
 			txrx = 0;
 
-			if (txbuf) {
-				txrx = ((uint8_t *)txbuf)[data->tx_count];
+			switch (data->dfs) {
+#if defined(CONFIG_SPI_RPI_PICO_PIO_32_BIT)
+			case 4: {
+				if (txbuf) {
+					txrx = ((uint32_t *)txbuf)[data->tx_count];
+				}
+				spi_pico_pio_sm_put32(data->pio, data->pio_sm, txrx);
+				data->tx_count += 4;
+			} break;
+#endif /* CONFIG_SPI_RPI_PICO_PIO_32_BIT */
+
+#if defined(CONFIG_SPI_RPI_PICO_PIO_16_BIT)
+			case 2: {
+				if (txbuf) {
+					txrx = ((uint16_t *)txbuf)[data->tx_count];
+				}
+				spi_pico_pio_sm_put16(data->pio, data->pio_sm, txrx);
+				data->tx_count += 2;
+			} break;
+#endif /* CONFIG_SPI_RPI_PICO_PIO_16_BIT */
+
+#if defined(CONFIG_SPI_RPI_PICO_PIO_8_BIT)
+			case 1: {
+				if (txbuf) {
+					txrx = ((uint8_t *)txbuf)[data->tx_count];
+				}
+				spi_pico_pio_sm_put8(data->pio, data->pio_sm, txrx);
+				data->tx_count++;
+			} break;
+#endif /*CONFIG_SPI_RPI_PICO_PIO_8_BIT*/
```

Comment:
This is just a matter of style and preference, but there are ifdefs all over the place, so it would be nice to see them put together. For example, how about changing the function to something like this?

```
static inline int spi_pico_pio_sm_put(PIO pio, uint sm, void *dataptr, size_t bits)
{
        switch(bits) {
		case 4:
			*((io_rw_32 *)&pio->txf[sm]) = *(uint32_t*)dataptr;
			return 4;
		case 2:
			*((io_rw_8 *)&pio->txf[sm]) = *(uint16_t*)dataptr;
			return 2;
		case 1:
			*((io_rw_8 *)&pio->txf[sm]) = *(uint8_t*)dataptr;
			return 1;
		default:
			return 0;
        }
}
```

Reply:
Unless one wants to save some ROM bytes, in this case the defines are useful. (unless, again, these defines disappear and become DTS properties)

Reply:
Condensing the ..._put and ..._get code, as @soburi suggested, is a good idea.  That makes the source code a lot shorter and cleaner.  I'll try to get that to work.